text/0000-read-buf.md

```diff
@@ -0,0 +1,626 @@
+both the standard library and the ecosystem today already works with uninitialized buffers because there are concrete,
+nontrivial performance improvements from doing so:
+
+* [The standard library measured](https://github.com/rust-lang/rust/pull/26950) a 7% improvement in benchmarks all the
+    way back in 2015.
+* [The hyper HTTP library measured](https://github.com/tokio-rs/tokio/pull/1744#issuecomment-554543881) a nontrivial
+    improvement in benchmarks.
+* [The Quinn QUIC library measured](https://github.com/tokio-rs/tokio/pull/1744#issuecomment-553501198) a 0.2%-2.45%
+    improvement in benchmarks.
+
+Given that the ecosystem has already found that uninitialized buffer use is important enough to deal with, the standard
+library should provide a more robust framework to work with.
+
+In addition, working with regular initialized buffers can be *more complex* than working with uninitialized buffers!
+Back in 2015, the standard library's implementation of `Read::read_to_end` was found to be wildly inefficient due to
+insufficiently careful management of buffer sizes because it was initializing them.
+[The fix](https://github.com/rust-lang/rust/pull/23820) improved the performance of small reads by over 4,000x! If
+the buffer did not need to be initialized, the simpler implementation would have been fine.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `ReadBuf` type manages a *progressively initialized* buffer of bytes. It is primarily used to avoid buffer
+initialization overhead when working with types implementing the `Read` trait. It wraps a buffer of
+possibly-uninitialized bytes and tracks how much of the buffer has been initialized and how much of the buffer has been
+filled. Tracking the set of initialized bytes allows initialization costs to only be paid once, even if the buffer is
+used repeatedly in a loop.
+
+Here's a small example of working with a reader using a `ReadBuf`:
+
+```rust
+// The base level buffer uses the `MaybeUninit` type to avoid having to initialize the whole 8kb of memory up-front.
+let mut buf = [MaybeUninit::<u8>::uninit(); 8192];
+
+// We then wrap that in a `ReadBuf` to track the state of the buffer.
+let mut buf = ReadBuf::uninit(&mut buf);
+
+loop {
+    // Read some data into the buffer.
+    some_reader.read_buf(&mut buf)?;
+
+    // If nothing was written into the buffer, we're at EOF.
+    if buf.filled().is_empty() {
+        break;
+    }
+
+    // Otherwise, process the data.
+    process_data(buf.filled());
+
+    // And then clear the buffer out so we can read into it again. This just resets the amount of filled data to 0,
+    // but preserves the memory of how much of the buffer has been initialized.
+    buf.clear();
+}
+```
+
+It is important that we created the `ReadBuf` outside of the loop. If we instead created it in each loop iteration we
```

Comment:
>  It is important that we created the `ReadBuf` outside of the loop.

Could it be a potential footgun for users if it was created inside the loop? So if users missed reading the documentation and have buffer inside the loop, will it be undefined behavior or will it just panic?

Reply:
It should just drop old values prematurely, nothing different from forgetting to use `io::Result<usize>` right now. But this is a potential foot-gun. There is no way we can panic or cause UB, because `read_buf` implementations can't rely on the same buffer being passed in each time.

Reply:
Values aren't being dropped, just the knowledge of how much of the buffer has been initialized in a previous iteration.

Reply:
So do we need to have something like `#[must_use]`? But still it is safe and no issues even though it is a potential foot-gun right?

Reply:
Yes, this is a safe API.

Reply:
Is there any way to prevent users from creating `ReadBuf` in a loop to prevent this footgun? Not all users read documentation though.